### PR TITLE
PUBDEV-8087 custom escape character

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-8087

### CODE REVIEW

File: h2o-core/src/main/java/water/api/schemas3/ParseV3.java [hunk 1]
```diff
@@ -69,4 +70,7 @@
   // Zero if blocking==false; row-count if blocking==true
   @API(help="Rows", direction=API.Direction.OUTPUT)
   public long rows;
+
+  @API(help="One-character string used to escape other characters.", direction=API.Direction.INOUT)
+  public byte escape_char = ParseSetup.DEFAULT_ESCAPE_CHAR;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/schemas3/ParseV3.java:75]: Should we actually make it a String in the API and then validate its properties?

File: h2o-py/h2o/frame.py [hunk 2]
```diff
@@ -445,28 +445,29 @@ def type(self, col):
 
 
     def _import_parse(self, path, pattern, destination_frame, header, separator, column_names, column_types, na_strings,
-                      skipped_columns=None, custom_non_data_line_markers=None, partition_by=None, quotechar=None):
+                      skipped_columns=None, custom_non_data_line_markers=None, partition_by=None, quotechar=None, escape_char=None):
```
Comment (michalkurka) [inline h2o-py/h2o/frame.py:448]: Should probably be consistent with naming of `quotechar`, Pandas also uses spelling `escapechar`

**Comment by michalkurka** [comment]
@sebhrusen can you please also take a look?

**Comment by koniecsveta** [comment]
R client in progress

File: h2o-core/src/main/java/water/api/schemas3/ParseV3.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/water/api/schemas3/ParseV3.java:75]: can be, I did it like byte since separator is also byte

Comment (michalkurka) [reply h2o-core/src/main/java/water/api/schemas3/ParseV3.java:75]: then it is probably better to be consistent, was just thinking out loud

File: h2o-py/h2o/h2o.py [hunk 3]
```diff
@@ -766,6 +770,7 @@ def parse_setup(raw_frames, destination_frame=None, header=0, separator=None, co
     :param custom_non_data_line_markers: If a line in imported file starts with any character in given string it will NOT be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
     :param partition_by: A list of columns the dataset has been partitioned by. None by default.
     :param quotechar: A hint for the parser which character to expect as quoting character. Only single quote, double quote or None (default) are allowed. None means automatic detection.
+    :param escape_char: One-character string used to escape other characters. '\' by default.
```
Comment (sebhrusen) [inline h2o-py/h2o/h2o.py:773]: Now that this is exposed, I'm wondering if it should not be None by default (char=0 on Java side) to be in sync with pandas.

Comment (koniecsveta) [reply h2o-py/h2o/h2o.py:773]: can be, but then there will be needed changes everywhere in tests where we are parsing something escaped by the escapechar which was hardcoded in h2o until now ('\\')

**Comment by koniecsveta** [comment]
Flow in progress

**Review by sebhrusen — COMMENTED** [review]
@koniecsveta the changes look good.

However, as mentioned earlier, for both internal consistency (the `escapechar=None` in Py is a bit confusing as it actually means `escapechar='\\'`, so how do we disable escaping?), and compatibility with pandas (for which escaping is disabled by default), I would disable escaping by default.
Which just means use `DEFAULT_ESCAPE_CHAR = 0`, and as you said, it may just require to explicitely set `escapechar='\\'` on a couple of existing tests, but that's fine as it was a very recent change.

File: h2o-py/h2o/h2o.py (hunk 3 above)
Comment (sebhrusen) [reply h2o-py/h2o/h2o.py]: correct, but it's only 2 or 3 of them: the ones I created in this PR: https://github.com/h2oai/h2o-3/pull/5388/files

**Comment by koniecsveta** [comment]
flow: https://github.com/h2oai/h2o-flow/pull/173

File: h2o-core/src/main/java/water/api/schemas3/ParseV3.java [hunk 4]
```diff
@@ -69,4 +70,7 @@
   // Zero if blocking==false; row-count if blocking==true
   @API(help="Rows", direction=API.Direction.OUTPUT)
   public long rows;
+
+  @API(help="One-character string used to escape other characters.", direction=API.Direction.INOUT)
+  public byte escapechar = ParseSetup.DEFAULT_ESCAPE_CHAR;
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/api/schemas3/ParseV3.java:75]: `byte` at API level? I'm amazed that it works.
In that case, maybe help should be updated to "One ASCII character used to..."

**Review by sebhrusen — APPROVED** [review]

File: h2o-core/src/main/java/water/api/schemas3/ParseV3.java (hunk 4 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/water/api/schemas3/ParseV3.java:75]: we have also `public byte separator`, so did it in the same way / same validation

**Pull request merged**